### ISSUE TITLE
Separate user and third party plugins

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot-proposals#2091]
### Describe the project you are working on
Plugins

### Describe the problem or limitation you are having in your project
Currently all plugins, whether user created or added from the asset lib, are installed in the addons folder.

Problems:
1) It's not possible to determine which plugins are third-party, and which are user ones from outside (for example, from a plugin/script) in order to do something with them

eg:

- a plugin could add a checkbox in the `FileSystem` dock that would hide the third-party addons folder
this way the user could not accidentally move files there, etc.

- a plugin could set a different icon/color for all third party scripts in the script list so that the user can more easily distinguish and not confuse scripts

- a plugin could (optionally) set `readonly` for third-party scripts in the code editor, so third-party scripts would not be available.
This would prevent accidental modification of third party scripts by the user in Godot's built-in script editor.

- a plugin could prohibit opening third-party scripts in the built-in script editor.

- a plugin could create a separate tab to display a list of third-party scripts in the built-in script editor

etc.

2) another problem is cluttering the user workspace with third-party scripts
It becomes more difficult for the user to find the folder with his plugin in the addons folder.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Separating user and third party plugins/addons seems logical to solve these problems.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

We could have two folders:
(a)
`res://addons` - for user addons
`res://third party addons` - for third party addons

`third party addons` looks somewhat lengthy, so a different name might be chosen.

or
(b)
`res://addons` - for third party addons
`res://user addons` - for user addons

Another option is:
(c)
`res://addons` - for third-party addons
`res://addons/user addons` - for user addons

or vice versa:
(d)
`res://addons` - for custom addons
`res://addons/third party` - for third party addons

Finally, there might be `user` and `third party` subfolders in the `addons` folder:
(e)
`res://addons/user` - for custom addons
`res://addons/third party` - for third party addons

---
Addons from asset libs would be installed in `third-party addons` folder;

`Create a Plugin` dialog should (at least by default) create plugins in `user addons` and so on

---
I'd like to note that using paths is not the only way to separate plugins into custom and third party: for this, for example, a config could be used (`plugin.cfg` could have information about whether the plugin is custom or third-party), but I think this would not solve problem (2).

### If this enhancement will not be used often, can it be worked around with a few lines of script?
plugin structure related

### Is there a reason why this should be core and not an add-on in the asset library?
plugin structure related

**Comment by pycbouh** [comment]
There is currently no practical difference between the two, as assets are just copied verbatim into your project. Even if mentally you consider some code third-party and some not, in practice everything is just a part of your codebase and behaves as a whole.

So maybe this should be a question presented for a more robust dependency system or a part of the global plugins solution. Not sure how much of the use it would be to just split folders now.

**Comment by me2beats** [comment]
I find this separation useful for the following reason:

There are scripts that the user creates and edits (I call these `user` scripts and plugins).
And there are scripts that are installed from `AssetLib`, or from other sources, editing which can lead to unpredictable things.

So I would like to make a plugin that does not allow editing plugin scripts from AssetLib and other third party plugins) by setting the `read_only` mode in the script editor.

And for this I need to know which scripts are "user" and which are "third-party" (installed from AssetLib, etc.).
But now there is no way to find out, because all plugins are in `addons`.

That's why I see the practical benefit of such a script separation.
it also somewhat solves the problem (2) described in the proposal.

I think the implementation of this one would require quite a few changes compared to the implementation of the dependency system, that will take perhaps a year or several years, which means it would be a practical workaround for now.

**Comment by pycbouh** [comment]
> editing which can lead to unpredictable things

True, but at the same time you may want to edit those third-party plugins, and you may want to disable editing of your own plugins if you've done building them and are only using them (or if you work in a team, and other members shouldn't be able to mess up the plugin code). 

So I think a read-only option would be a good thing, I'd even suggest it could be made a built-in thing, instead of you make it a plugin, but I don't think that it should work on the basis of third-party vs first-party.

**Comment by me2beats** [comment]
@pycbouh But how could Godot figure out which script is readonly then?
`_readonly()` virtual method?

**Comment by pycbouh** [comment]
No, I imagine it being a configuration option, no code involved. Probably a project setting with a list of readonly plugins would be a good idea. Same as we have a list of enabled plugins.

**Comment by me2beats** [comment]
Ok, it looks like the implementation of things like `readonly` shouldn't be hardcoded to dividing plugins into user and third-party only, 
But what do you think about problem (2)?

Imo "user addons together with third-party ones" clutters up the user workspace.

**Comment by pycbouh** [comment]
I haven't really hit that amount of addons to experience what you describe, but something similar happens with Node.js, Composer and python modules and is annoying. However, I think this is inevitable and should be solved by better editor tools for accessing addons (with search, for example), and by global addons which are installed per system user and are enabled by default for all projects, even new projects. Half of my addons are not required for individual projects and would happily be relocated to a user folder.

**Comment by dalexeev** [comment]
I suggest the `source[/author]/name` scheme. Examples:
`res://addons/localhost/my_tools`
`res://addons/github.com/lawnjelly/smoothing-addon`

Although, it is best for plugins to use relative paths everywhere and don't care what `res://addons/folder` they are installed in, and the user can rename `folder`.

**Comment by fedetibaldo** [comment]
I could also see it implemented as a project setting that lets you add more "plugin root folders". This way:
- if I want to rename the third party addons folder to `vendor` (after updating the relevant plugin code), I can
- if I want to create a separate `user_addons` folder, I can
- if I want to register a plugin inside my `core/feature_xyz/editor` folder, I can too
